### ISSUE TITLE
Compiler doesn't warn on infinite recursion in trait's method

### ISSUE THREAD

**Issue by DoumanAsh** [rust-lang/rust#78521]
Example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4782b3bebc1e92369e5ff7b3c5b4a6e1

Visible on both stable and nightly.

Compiler warns on infinite recursion in case of regular method:

```rust
impl AsyncTokioTimer<RawTimer> {
    #[inline]
    ///Creates new instance
    pub fn new(_time: time::Duration) -> Self {
        Self::new(_time)
    }
}
```

Compiler doesn't issue the same warning in case of trait's method

```rust
impl<T: TimerFd> Timer for AsyncTokioTimer<T> {
    fn new(timeout: time::Duration) -> Self {
        debug_assert!(timeout.as_millis() <= u32::max_value().into());
        Self::new(timeout)
    }
}
```

I suspect it should?

**Comment by jonas-schievink** [comment]
It's because the `debug_assert!` *might* unwind instead of falling through to the recursion below

**Comment by SNCPlay42** [comment]
The `Self::new` in the trait impl there refers to the inherent `new`, not the one from the trait impl.

e.g. this code does not recurse infinitely: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c31a36d86be2dfbb4a9702325d1fb36c

**Comment by DoumanAsh** [comment]
@jonas-schievink Ah you're correct, although to be honest I believe it should still warn, as otherwise you get abort at runtime without even knowing reason (since it is just stack overflow)

@SNCPlay42 in my case it does call, which leads to stack overflow otherwise I wouldn't care issue

**Comment by jyn514** [comment]
Seems like a duplicate of https://github.com/rust-lang/rust/issues/78474.

**Comment by jonas-schievink** [comment]
It does – but this issue actually invokes a diverging function, so the proposed fix for that issue in https://github.com/rust-lang/rust/issues/78474#issuecomment-717879443 probably won't fix this one

**Comment by keiichiw** [comment]
I think the idea of https://github.com/rust-lang/rust/issues/78474#issuecomment-717879443 can fix it because the return type of `debug_assert!` is `()` not `!`. In my understanding, we say "a function is diverging" only if the function *never* return a value.

**Comment by DoumanAsh** [comment]
`debug_assert!` should expands to optional panic which is `!`

**Comment by keiichiw** [comment]
I meant `debug_assert!` can seen as the same with the following function, whose return type is `()`.
```rust
fn debug_assert_func(flg : bool) -> () {
  if !flg {
    panic!()
  }
}
```

Ah, but, it's a macro not a function. So, the type-based approach in #78474 may not be applicable here...
